File: packages/expo-conductor/android/src/main/java/expo/modules/conductor/engine/Policy.kt
package expo.modules.conductor.engine

/**
 * Policy / constraint evaluation engine (Android). Mirrors `web/engine/policy.ts`.
 * Checks run in a fixed order; first failure wins. Validated by
 * `fixtures/policy.cases.json`.
 */
object PolicyEngine {
  fun evaluate(constraints: Constraints, ctx: DeviceContext): PolicyDecision {
    val now = ctx.now

    constraints.expiresAt?.let { if (now > it) return fail(PolicyReason.EXPIRED) }
    constraints.window?.earliest?.let { if (now < it) return fail(PolicyReason.BEFORE_WINDOW) }
    constraints.window?.latest?.let { if (now > it) return fail(PolicyReason.AFTER_WINDOW) }
    if (constraints.requiresCharging == true && !ctx.charging) {
      return fail(PolicyReason.REQUIRES_CHARGING)
    }
    constraints.minBatteryLevel?.let {
      if (ctx.batteryLevel < it) return fail(PolicyReason.BATTERY_TOO_LOW)
    }
    if (constraints.network == "any" && ctx.networkType == "none") {
      return fail(PolicyReason.NETWORK_UNAVAILABLE)
    }
    if (constraints.network == "unmetered" && ctx.networkType != "unmetered") {
      return fail(PolicyReason.NETWORK_NOT_UNMETERED)
    }
    if (constraints.requiresIdle == true && !ctx.idle) {
      return fail(PolicyReason.REQUIRES_IDLE)
    }
    return PolicyDecision(true, PolicyReason.ELIGIBLE)
  }

  private fun fail(reason: PolicyReason) = PolicyDecision(false, reason)
}

File: packages/expo-conductor/ios/Engine/Policy.swift
import Foundation

/// Policy / constraint evaluation engine (iOS). Mirrors `web/engine/policy.ts`.
/// Checks run in a fixed order; first failure wins. Validated by
/// `fixtures/policy.cases.json`.
public enum PolicyEngine {
  public static func evaluate(_ constraints: Constraints, _ ctx: DeviceContext) -> PolicyDecision {
    let now = ctx.now

    if let expiresAt = constraints.expiresAt, now > expiresAt {
      return fail(.expired)
    }
    if let earliest = constraints.window?.earliest, now < earliest {
      return fail(.beforeWindow)
    }
    if let latest = constraints.window?.latest, now > latest {
      return fail(.afterWindow)
    }
    if constraints.requiresCharging == true, !ctx.charging {
      return fail(.requiresCharging)
    }
    if let minBattery = constraints.minBatteryLevel, ctx.batteryLevel < minBattery {
      return fail(.batteryTooLow)
    }
    if constraints.network == "any", ctx.networkType == "none" {
      return fail(.networkUnavailable)
    }
    if constraints.network == "unmetered", ctx.networkType != "unmetered" {
      return fail(.networkNotUnmetered)
    }
    if constraints.requiresIdle == true, !ctx.idle {
      return fail(.requiresIdle)
    }
    return PolicyDecision(eligible: true, reason: .eligible)
  }

  private static func fail(_ reason: PolicyReason) -> PolicyDecision {
    PolicyDecision(eligible: false, reason: reason)
  }
}
